File: code/client/core/source/value-table/ValueTableManager.cpp
// <copyright file="ValueTableManager.cpp" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

module;

#include <memory>
#include <stdexcept>

export module Soup.Core:ValueTableReader;

import Opal;
import :Value;
import :ValueTableReader;
import :ValueTableWriter;

using namespace Opal;

namespace Soup::Core
{
	/// <summary>
	/// The Value Table state manager
	/// </summary>
	export class ValueTableManager
	{
	public:
		/// <summary>
		/// Load the value table from the target file
		/// </summary>
		static bool TryLoadState(
			const Path& valueTableFile,
			ValueTable& result)
		{
			// Open the file to read from
			std::shared_ptr<System::IInputFile> file;
			if (!System::IFileSystem::Current().TryOpenRead(valueTableFile, true, file))
			{
				Log::Info("Value Table file does not exist");
				return false;
			}

			// Read the contents of the build state file
			try
			{
				result = ValueTableReader::Deserialize(file->GetInStream());
				return true;
			}
			catch(std::runtime_error& ex)
			{
				Log::Error(ex.what());
				return false;
			}
			catch(...)
			{
				Log::Error("Failed to parse value table");
				return false;
			}
		}

		/// <summary>
		/// Save the value table for the target file
		/// </summary>
		static void SaveState(
			const Path& valueTableFile,
			ValueTable& state)
		{
			auto targetFolder = valueTableFile.GetParent();

			// Open the file to write to
			auto file = System::IFileSystem::Current().OpenWrite(valueTableFile, true);

			// Write the build state to the file stream
			ValueTableWriter::Serialize(state, file->GetOutStream());
		}
	};
}

File: code/generate-sharp/utilities/value-table/ValueTableManager.cs
﻿// <copyright file="ValueTableManager.cs" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

using Opal;
using Opal.System;
using System;
using System.Text;
using Path = Opal.Path;

namespace Soup.Build.Utilities;

/// <summary>
/// The Value Table state manager
/// </summary>
public static class ValueTableManager
{
	/// <summary>
	/// Load the value table from the target file
	/// </summary>
	public static bool TryLoadState(
		Path valueTableFile,
		out ValueTable result)
	{
		// Verify the requested file exists
		if (!LifetimeManager.Get<IFileSystem>().Exists(valueTableFile))
		{
			Log.Info("Value Table file does not exist");
			result = [];
			return false;
		}

		// Open the file to read from
		using var file = LifetimeManager.Get<IFileSystem>().OpenRead(valueTableFile);
		using var reader = new System.IO.BinaryReader(file.GetInStream(), Encoding.UTF8, true);

		// Read the contents of the build state file
		try
		{
			result = ValueTableReader.Deserialize(reader);
			return true;
		}
		catch (InvalidOperationException ex)
		{
			Log.Error($"Failed to parse value table: {ex.Message}");
			result = [];
			return false;
		}
	}

	/// <summary>
	/// Save the value table for the target file
	/// </summary>
	public static void SaveState(
		Path valueTableFile,
		ValueTable state)
	{
		// Open the file to write to
		using var fileStream = System.IO.File.Open(valueTableFile.ToString(), System.IO.FileMode.Create, System.IO.FileAccess.Write);
		using var writer = new System.IO.BinaryWriter(fileStream);

		// Write the build state to the file stream
		ValueTableWriter.Serialize(state, writer);
	}
}